Replace Laminas\Crypt library

demiankatz opened the pull request:
This PR replaces the abandoned laminas-crypt dependency with a new local class containing equivalent logic derived (and dramatically simplified) from the behavior of that library.

In several cases, laminas-crypt mechanisms have been replaced with native PHP equivalents that did not exist when laminas-crypt was originally developed. However, there is one problem area: the laminas-crypt implementation of pbkdf2 produces results that are incompatible with the native PHP implementation of the same algorithm. For now, this PR is coded to always use native PHP for encryption, but to fall back to the legacy Laminas algorithm for decryption for back-compatibility. We should discuss a strategy for converting legacy data so that we can eventually deprecate and remove this legacy logic.

TODO
- [x] Refactor Laminas\Crypt logic to a central place
- [x] Replace Laminas\Crypt logic
- [x] Test everything
- [x] Discussion: do we need tooling to re-encrypt legacy data so we can deprecate legacy algorithms? (Or should we advise using existing tooling to regenerate?)
- [x] Discussion: should \VuFind\Crypt\BlockCipher become a top-level service (or at least have a top-level service that serves as a factory for BlockCipher objects) for easier extensibility?
- [x] Resolve [VUFIND-1705](https://openlibraryfoundation.atlassian.net/issues/VUFIND-1705) when merging
- [x] Update changelog when merging (note removal of library, relevant constructor signature changes: VuFind\Auth\ILSAuthenticator, VuFind\Session\SecureDelegator, VuFindConsole\Command\SwitchDbHashCommand, and any best-practice data conversion suggestions)
- [x] After merging, open new JIRA ticket to change the legacyPbkdf2 behavior in release 12.0.

demiankatz committed:
Refactor encryption to a central class.

demiankatz committed:
Refactor code.

demiankatz committed:
Refactor laminas-crypt code into self-contained class.

demiankatz committed:
Remove laminas-crypt dependency.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into vufind-1705

demiankatz committed:
Remove Laminas\Rand dependency.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into vufind-1705

demiankatz committed:
Clean up exceptions; use native hash_pbkdf2 where possible.

demiankatz committed:
Style fix.

demiankatz committed:
Normalize whitespace.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+    protected function setAlgorithm(string $algo): static
+    {
+        if (!in_array($this->encryptionAlgos[$algo] . '-cbc', openssl_get_cipher_methods(true))) {
+            throw new InvalidArgumentException(
```
This would fit on a single line.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+     */
+    protected function setAlgorithm(string $algo): static
+    {
+        if (!in_array($this->encryptionAlgos[$algo] . '-cbc', openssl_get_cipher_methods(true))) {
```
This should check that `$this->encryptionAlgos` contains `$algo` first.

EreMaijala replied:
Should this take `$this->mode` into account instead of hardcoded `-cbc`?

demiankatz replied:
The original Laminas code had a hard-coded cbc (possibly as a "least common denominator"), but I've revised our version to take mode into consideration via a new setAlgorithmAndMode() method that gets wrapped by setAlgorithm(). I think this is safe for now, and it may make it easier to introduce new options in future if the need arises.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+     */
+    public function encrypt(string $data): string
+    {
+        // 0 (as integer), 0.0 (as float) & '0' (as string) will return false, though these should be allowed
```
I think I understand what this refers to, but it still feels oddly out of place. Perhaps something like `// Compare against an empty string, because...` ?

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+        $ciphertext = base64_decode(mb_substr($data, $hmacSize, null, '8bit') ?: '');
+        $iv = mb_substr($ciphertext, 0, $this->getSaltSize(), '8bit');
+        $hmacNew = $this->getValidationHmac($ciphertext, $iv, $keySize);
+        if (strcmp($hmacNew, $hmac) !== 0) {
```
Couldn't this be just `$hmacNew !== $hmac`? If not, a comment would be useful.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+        if (strcmp($hmacNew, $hmac) !== 0) {
+            // If authentication failed using new algorithm, fall back to legacy:
+            $hmacNew = $this->getValidationHmac($ciphertext, $iv, $keySize, true);
+            if (strcmp($hmacNew, $hmac) !== 0) {
```
Same as above.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+     * @return string
+     * @throws InvalidArgumentException
+     */
+    protected function getValidationHmac(string $ciphertext, string $salt, int $keySize, bool $legacy = false): string
```
I don't really like it that this sets the OpenSSL key kind of like a side effect without anything in the function name implying it.

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,550 @@
+    }
+
+    /**
+     * Generate the encryption key and the HMAC key for the authentication
```
encryption/decryption?

EreMaijala commented on module/VuFindConsole/src/VuFindConsole/Command/Util/SwitchDbHashCommand.php:
```diff
@@ -195,8 +194,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
         // Initialize Openssl first, so we can catch any illegal algorithms before
```
This isn't really true anymore. But we should initialize the BlockCipher instances before changing configuration further below to catch any invalid algorithm.

EreMaijala commented on module/VuFindConsole/src/VuFindConsole/Command/Util/SwitchDbHashCommand.php:
```diff
@@ -195,8 +194,8 @@ protected function execute(InputInterface $input, OutputInterface $output)
         try {
-            $oldCrypt = $this->getOpenSsl($oldhash);
-            $newCrypt = $this->getOpenSsl($newhash);
+            $oldCrypt = $this->getCipherOptions($oldhash);
```
These variables are pretty useless, but perhaps create the BlockCipher instances here instead?

demiankatz committed:
Merge branch 'dev' into vufind-1705

demiankatz committed:
Review feedback.

demiankatz commented:
Thanks, @EreMaijala! Commit [a9a375f](https://github.com/vufind-org/vufind/pull/4089/commits/a9a375f6cde76ff2156a70053d7f745a1202b3ea) should address all of your feedback. I'm currently running Mink tests locally but I'm pretty confident they should still pass. I'll report back in a few minutes with the results.

demiankatz commented:
Tests are passing!

EreMaijala commented on module/VuFind/src/VuFind/Crypt/BlockCipher.php:
```diff
@@ -0,0 +1,568 @@
+     * @return string
+     * @throws InvalidArgumentException
+     */
+    protected function generateKeyandReturnValidationHmac(
```
Wrong case in `and`, should be `generateKeyAndReturnValidationHmac`

demiankatz replied:
The curse of copy-and-paste; oops! Fixed now. :-)

demiankatz committed:
Fix typo in method name.

EreMaijala reviewed (commented):
The code now looks good to me, but I suddenly realized that this has the potential to cause trouble if e.g. someone finds that they need to downgrade an already-upgraded installation, or needs to work with several different versions side-by-side (development or migration testing etc.). So I think we should default to using the legacy Pbkdf2 algorithm and only start using the `hash_pbkdf2` function if explicitly enabled.

demiankatz committed:
Merge branch 'dev' into vufind-1705

demiankatz committed:
Refactor and adjust comments/names for clarity.

demiankatz committed:
Add legacy config setting; refactor to use service.

demiankatz committed:
Merge branch 'dev' into vufind-1705

demiankatz commented:
Thanks, @EreMaijala, I've added your suggested setting, which required a bit more refactoring and the introduction of BlockCipher as a service (which I was considering doing anyway, but this forced it). I think this is ready for another look! Mink tests are passing, and I've done manual testing to confirm that legacy setting fallback works correctly in both directions.

EreMaijala reviewed (approved):
Looks good to me now. 

If you want, you can change the new `callable`'s in constructors to `Closure` and use `Closure::fromCallable()` to enable constructor property promotion for them, but it's fine for me if you prefer to keep them `callable`.

demiankatz committed:
Use Closure instead of callable.

demiankatz commented:
Good idea, @EreMaijala! I've gone ahead and switched callables to Closures in commit [5833855](https://github.com/vufind-org/vufind/pull/4089/commits/5833855845471bc7bd560801e81bc67e63884748). I also modified the existing callable in the ILSAuthenticator constructor while I was changing the signature anyway.

demiankatz commented:
I've also run the Mink tests again, and also done some manual testing to confirm that running util/switch_db_hash correctly updates everything when you change the legacy setting in config.ini. I think this is all set, but I'll wait for one more approval of my latest changes, just in case I missed anything.
